### ISSUE TITLE
Rails 7 persistent storage not working

### ISSUE THREAD

**Issue by mx-tiem** [dokku/dokku#5260]
### Description of problem

I'm new to dokku and I'm having a great time with it, but I can't make persistent storage to work. My uploaded images get lost after each restart or deploy.

I was following this tutorial https://www.pedroalonso.net/blog/hosting-rails-in-dokku

The only problem with that tutorial is that I couldn't chown storage folder until I added dokku user to sudoers

Let me know if I can provide you anymore info, but at this point I don't know where to look anymore.

### Steps to reproduce

https://www.pedroalonso.net/blog/hosting-rails-in-dokku

### dokku report $APP_NAME

-----> uname: Linux xxxx.l.timexxxx.xxxxcloud 5.4.0-28-generic #32-Ubuntu SMP Wed Apr 22 17:40:10 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
-----> memory: 
                     total        used        free      shared  buff/cache   available
       Mem:           1987         681         429          37         876        1079
       Swap:             0           0           0
-----> docker version: 
       Client: Docker Engine - Community
        Version:           20.10.17
        API version:       1.41
        Go version:        go1.17.11
        Git commit:        100c701
        Built:             Mon Jun  6 23:02:57 2022
        OS/Arch:           linux/amd64
        Context:           default
        Experimental:      true
       
       Server: Docker Engine - Community
        Engine:
         Version:          20.10.17
         API version:      1.41 (minimum version 1.12)
         Go version:       go1.17.11
         Git commit:       a89b842
         Built:            Mon Jun  6 23:01:03 2022
         OS/Arch:          linux/amd64
         Experimental:     false
        containerd:
         Version:          1.6.6
         GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
        runc:
         Version:          1.1.2
         GitCommit:        v1.1.2-0-ga916309
        docker-init:
         Version:          0.19.0
         GitCommit:        de40ad0
-----> docker daemon info: 
WARNING: No swap limit support
       Client:
        Context:    default
        Debug Mode: true
        Plugins:
         app: Docker App (Docker Inc., v0.9.1-beta3)
         buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
         compose: Docker Compose (Docker Inc., v2.6.0)
         scan: Docker Scan (Docker Inc., v0.17.0)
       
       Server:
        Containers: 8
         Running: 6
         Paused: 0
         Stopped: 2
        Images: 35
        Server Version: 20.10.17
        Storage Driver: overlay2
         Backing Filesystem: extfs
         Supports d_type: true
         Native Overlay Diff: true
         userxattr: false
        Logging Driver: json-file
        Cgroup Driver: cgroupfs
        Cgroup Version: 1
        Plugins:
         Volume: local
         Network: bridge host ipvlan macvlan null overlay
         Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
        Swarm: inactive
        Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
        Default Runtime: runc
        Init Binary: docker-init
        containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
        runc version: v1.1.2-0-ga916309
        init version: de40ad0
        Security Options:
         apparmor
         seccomp
          Profile: default
        Kernel Version: 5.4.0-28-generic
        Operating System: Ubuntu 20.04 LTS
        OSType: linux
        Architecture: x86_64
        CPUs: 1
        Total Memory: 1.941GiB
        Name: xxxx.l.time4xxxxx.xxxcloud
        ID: GJQI:R4FX:QGUP:JC53:E77M:I3ZX:EG42:LLBA:5TF6:22CN:4ZZT:VY7L
        Docker Root Dir: /var/lib/docker
        Debug Mode: false
        Registry: https://index.docker.io/v1/
        Labels:
        Experimental: false
        Insecure Registries:
         127.0.0.0/8
        Live Restore Enabled: false
       
-----> git version: git version 2.25.1
-----> sigil version: 0.9.0build+bc921b7
-----> herokuish version: 
       herokuish: 0.5.36
       buildpacks:
         heroku-buildpack-multi     v1.2.0
         heroku-buildpack-ruby      v240
         heroku-buildpack-nodejs    v196
         heroku-buildpack-clojure   v87
         heroku-buildpack-python    v211
         heroku-buildpack-java      v70
         heroku-buildpack-gradle    v36
         heroku-buildpack-scala     v92
         heroku-buildpack-play      v26
         heroku-buildpack-php       v218
         heroku-buildpack-go        v162
         heroku-buildpack-nginx     v16
         buildpack-null             v3
-----> dokku version: dokku version 0.27.7
-----> plugn version: plugn: 0.12.0build+3a27594
-----> dokku plugins: 
         00_dokku-standard    0.27.7 enabled    dokku core standard plugin
         20_events            0.27.7 enabled    dokku core events logging plugin
         app-json             0.27.7 enabled    dokku core app-json plugin
         apps                 0.27.7 enabled    dokku core apps plugin
         builder              0.27.7 enabled    dokku core builder plugin
         builder-dockerfile   0.27.7 enabled    dokku core builder-dockerfile plugin
         builder-herokuish    0.27.7 enabled    dokku core builder-herokuish plugin
         builder-null         0.27.7 enabled    dokku core builder-null plugin
         builder-pack         0.27.7 enabled    dokku core builder-pack plugin
         buildpacks           0.27.7 enabled    dokku core buildpacks plugin
         certs                0.27.7 enabled    dokku core certificate management plugin
         checks               0.27.7 enabled    dokku core checks plugin
         common               0.27.7 enabled    dokku core common plugin
         config               0.27.7 enabled    dokku core config plugin
         cron                 0.27.7 enabled    dokku core cron plugin
         docker-options       0.27.7 enabled    dokku core docker-options plugin
         domains              0.27.7 enabled    dokku core domains plugin
         enter                0.27.7 enabled    dokku core enter plugin
         git                  0.27.7 enabled    dokku core git plugin
         letsencrypt          0.18.0 enabled    Automated installation of let's encrypt TLS certificates
         logs                 0.27.7 enabled    dokku core logs plugin
         network              0.27.7 enabled    dokku core network plugin
         nginx-vhosts         0.27.7 enabled    dokku core nginx-vhosts plugin
         plugin               0.27.7 enabled    dokku core plugin plugin
         postgres             1.22.0 enabled    dokku postgres service plugin
         proxy                0.27.7 enabled    dokku core proxy plugin
         ps                   0.27.7 enabled    dokku core ps plugin
         redis                1.22.0 enabled    dokku redis service plugin
         registry             0.27.7 enabled    dokku core registry plugin
         repo                 0.27.7 enabled    dokku core repo plugin
         resource             0.27.7 enabled    dokku core resource plugin
         run                  0.27.7 enabled    dokku core run plugin
         scheduler            0.27.7 enabled    dokku core scheduler plugin
         scheduler-docker-local 0.27.7 enabled    dokku core scheduler-docker-local plugin
         scheduler-null       0.27.7 enabled    dokku core scheduler-null plugin
         shell                0.27.7 enabled    dokku core shell plugin
         ssh-keys             0.27.7 enabled    dokku core ssh-keys plugin
         storage              0.27.7 enabled    dokku core storage plugin
         trace                0.27.7 enabled    dokku core trace plugin
=====> bobo-servis app-json information
       App json computed selected:    app.json
       App json global selected:      app.json
       App json selected:             
=====> bobo-servis app information
       App created at:                1658320299
       App deploy source:             bobo-servis
       App deploy source metadata:    bobo-servis
       App dir:                       /home/dokku/bobo-servis
       App locked:                    false
=====> bobo-servis builder information
       Builder build dir:             
       Builder computed build dir:    
       Builder computed selected:     
       Builder global build dir:      
       Builder global selected:       
       Builder selected:              
=====> bobo-servis builder-dockerfile information
       Builder dockerfile computed dockerfile path: Dockerfile               
       Builder dockerfile global dockerfile path: Dockerfile               
       Builder dockerfile dockerfile path:                          
=====> bobo-servis builder-pack information
       Builder pack computed projecttoml path: project.toml             
       Builder pack global projecttoml path: project.toml             
       Builder pack projecttoml path:                          
=====> bobo-servis buildpacks information
       Buildpacks computed stack:     gliderlabs/herokuish:latest-20
       Buildpacks global stack:       
       Buildpacks list:               
       Buildpacks stack:              
=====> bobo-servis ssl information
       Ssl dir:                       /home/dokku/bobo-servis/tls
       Ssl enabled:                   false                    
       Ssl hostnames:                                          
       Ssl expires at:                                         
       Ssl issuer:                                             
       Ssl starts at:                                          
       Ssl subject:                                            
       Ssl verified:                                           
=====> bobo-servis checks information
       Checks disabled list:          none                     
       Checks skipped list:           none                     
=====> bobo-servis cron information
       Cron task count:               0
=====> bobo-servis docker options information
       Docker options build:          --link dokku.postgres.bobo-servis-db:dokku-postgres-bobo-servis-db --link dokku.redis.bobo-servis-redis:dokku-redis-bobo-servis-redis 
       Docker options deploy:         --link dokku.postgres.bobo-servis-db:dokku-postgres-bobo-servis-db --link dokku.redis.bobo-servis-redis:dokku-redis-bobo-servis-redis --restart=on-failure:10 -v /var/lib/dokku/data/storage/bobo-servis-storage:/storage 
       Docker options run:            --link dokku.postgres.bobo-servis-db:dokku-postgres-bobo-servis-db --link dokku.redis.bobo-servis-redis:dokku-redis-bobo-servis-redis -v /var/lib/dokku/data/storage/bobo-servis-storage:/storage 
=====> bobo-servis domains information
       Domains app enabled:           true                     
       Domains app vhosts:            xx.xx.xx.xx              
       Domains global enabled:        true                     
       Domains global vhosts:         xx.xx.xx.xx              
=====> bobo-servis git information
       Git deploy branch:             master                   
       Git global deploy branch:      master                   
       Git keep git dir:              false                    
       Git rev env var:               GIT_REV                  
       Git sha:                       219d189                  
       Git last updated at:           1658316257               
Can't open /home/dokku/bobo-servis/tls/server.crt for reading, No such file or directory
139920407536960:error:02001002:system library:fopen:No such file or directory:../crypto/bio/bss_file.c:69:fopen('/home/dokku/bobo-servis/tls/server.crt','r')
139920407536960:error:2006D080:BIO routines:BIO_new_file:no such file:../crypto/bio/bss_file.c:76:
unable to load certificate
=====> bobo-servis letsencrypt information
       Letsencrypt active:            false                    
       Letsencrypt autorenew:         false                    
       Letsencrypt email:             xxxxx@xxxxx.com        
       Letsencrypt expiration:        1658275200               
=====> bobo-servis logs information
       Logs computed max size:        10m
       Logs global max size:          10m
       Logs global vector sink:       
       Logs max size:                 
       Logs vector sink:              
=====> bobo-servis network information
       Network attach post create:           
       Network attach post deploy:           
       Network bind all interfaces:          false
       Network computed attach post create:  
       Network computed attach post deploy:  
       Network computed bind all interfaces: false
       Network computed initial network:     
       Network computed tld:                 
       Network global attach post create:    
       Network global attach post deploy:    
       Network global bind all interfaces:   false
       Network global initial network:       
       Network global tld:                   
       Network initial network:              
       Network static web listener:          
       Network tld:                          
       Network web listeners:                172.17.0.6:5000
=====> bobo-servis nginx information
       Nginx access log format:                                
       Nginx access log path:         /var/log/nginx/bobo-servis-access.log
       Nginx bind address ipv4:                                
       Nginx bind address ipv6:       ::                       
       Nginx client max body size:                             
       Nginx disable custom config:   false                    
       Nginx error log path:          /var/log/nginx/bobo-servis-error.log
       Nginx global hsts:             true                     
       Nginx computed hsts:           true                     
       Nginx hsts:                                             
       Nginx hsts include subdomains: true                     
       Nginx hsts max age:            15724800                 
       Nginx hsts preload:            false                    
       Nginx proxy buffer size:       4096                     
       Nginx proxy buffering:         on                       
       Nginx proxy buffers:           8 4096                   
       Nginx proxy busy buffers size: 8192                     
       Nginx proxy read timeout:      60s                      
       Nginx last visited at:         1658320409               
       Nginx x forwarded for value:   $remote_addr             
       Nginx x forwarded port value:  $server_port             
       Nginx x forwarded proto value: $scheme                  
       Nginx x forwarded ssl:                                  
=====> bobo-servis proxy information
       Proxy enabled:                 true
       Proxy port map:                http:80:5000
       Proxy type:                    nginx
=====> bobo-servis ps information
       Deployed:                      true
       Processes:                     2
       Ps can scale:                  true
       Ps computed procfile path:     Procfile
       Ps global procfile path:       Procfile
       Ps procfile path:              
       Ps restart policy:             on-failure:10
       Restore:                       true
       Running:                       true
       Status web 1:                  running (CID: e5136dc98d5)
       Status worker 1:               running (CID: 6a117d8bc16)
=====> bobo-servis registry information
       Registry computed image repo:      dokku/bobo-servis
       Registry computed push on release: false
       Registry computed server:          
       Registry global push on release:   
       Registry global server:            
       Registry image repo:               
       Registry push on release:          
       Registry server:                   
       Registry tag version:              
=====> bobo-servis resource information
=====> bobo-servis scheduler information
       Scheduler computed selected:   docker-local
       Scheduler global selected:     docker-local
       Scheduler selected:            
=====> bobo-servis scheduler-docker-local information
       Scheduler docker local disable chown:                          
       Scheduler docker local parallel schedule count:                          
=====> bobo-servis storage information
       Storage build mounts:                                   
       Storage deploy mounts:         -v /var/lib/dokku/data/storage/bobo-servis-storage:/storage 
       Storage run mounts:            -v /var/lib/dokku/data/storage/bobo-servis-storage:/storage

**Comment by josegonzalez** [comment]
> The only problem with that tutorial is that I couldn't chown storage folder until I added dokku user to sudoers

What did you do here exactly?

**Comment by mx-tiem** [comment]
accidentally closed,sry

**Comment by josegonzalez** [comment]
Gotcha. You should be able to replace some (all?) of that with the `storage:ensure-directory` command, which takes an optional `--chown` flag. Docs are [here](https://dokku.com/docs/advanced-usage/persistent-storage/#creating-storage-directories). This should cover the request, as there aren't any extra perms needed to create that directory with the built-in command (though maybe that tutorial should be updated).

```
# create it with the perms
dokku storage:ensure-directory --chown herokuish rails7-sample-storage

# mount it
dokku storage:mount rails7-sample /var/lib/dokku/data/storage/rails7-sample-storage:/storage
```

Does that cover your use case?

**Comment by mx-tiem** [comment]
I did find today about ensure-directory with --chown, but I ran it for dokku, as:
```
dokku storage:ensure-directory --chown dokku rails7-sample-storage
```

I reran it now as you said with, herokuish user, and everything looked well, it said it's ready to mount.

When I tried to mount it, dokku said it already is mounted, so I unmounted and remounted it back and restarted the app (bobo-servis).

Opened app in browser and tried uploading an image, everything looked well, and then I restarted bobo-servis with:

```
dokku ps:restart bobo-servis
```

And the images were gone as usual. Here's ls -a of before and after uploading of an image
```
before upload:

dokku@44g7:~$ dokku enter bobo-servis web
herokuishuser@ad70412c360b:~$ ls -a storage
.  ..  .keep


dokku@44g7:~$ ls -a /var/lib/dokku/data/storage/bobo-servis-storage/
.  ..



after upload:

dokku@44g7:~$ dokku enter bobo-servis web
herokuishuser@ad70412c360b:~$ ls -a storage
.  ..  7u  .keep

dokku@44g7:~$ ls -a /var/lib/dokku/data/storage/bobo-servis-storage/
.  ..
```

I'll upload app report in a moment.

**Comment by josegonzalez** [comment]
You need to rebuild the app at least once so the mount is added. I don't think a restart will add it.

Can you post the output of `dokku ps:inspect $APP`, where `$APP` is your app name?

**Comment by josegonzalez** [comment]
Okay yeah it looks like the volume is mounted. Are you sure your app is writing to `/storage` and not `/app/storage`?

**Comment by mx-tiem** [comment]
Well, sure as I can be. It's a rails 7 app with vastly default settings and basic gems. You can find example of an simple rails app on:

https://github.com/pedropaf/rails7-sample-dokku

and here you can see that storage folder is in a root of an app

Also, I did try ls -a both in app and on server before and after an upload. The results:

```
before upload:

dokku@44g7:~$ dokku enter bobo-servis web
herokuishuser@ad70412c360b:~$ ls -a storage
.  ..  .keep


dokku@44g7:~$ ls -a /var/lib/dokku/data/storage/bobo-servis-storage/
.  ..



after upload:

dokku@44g7:~$ dokku enter bobo-servis web
herokuishuser@ad70412c360b:~$ ls -a storage
.  ..  7u  .keep

dokku@44g7:~$ ls -a /var/lib/dokku/data/storage/bobo-servis-storage/
.  ..
```

that '7u' file is stored photo, and those files went missing after each restart (as it should as I understood), but they should be backed up automatically if storage is correctly mounted

That does not happens as you can see on after upload snippet

Currently waiting on redeploy, that should rebuild it

**Comment by josegonzalez** [comment]
That path in the container is probably /app/storage. Paths are not mounted relative to the workdir, and the workdir for buildpack-based apps is `/app`.

**Comment by josegonzalez** [comment]
When you enter the container, run `echo $PWD` and you'll see the current working directory (which will be `/app`).

**Comment by mx-tiem** [comment]
ok, I think I got it :)

full path of app's storage folder is /app/storage

there's no home or user or anything. I got it confused with app folder in rails app. full path of rails app folder in dokku container is:
/app/app

I unmounted and remounted storage folder to track /app/storage (and deployed again)

**AND IT WORKED!**

also

Don't add herokuish or dokku to sudoers, it breaks up deployment bad. that's why late reply. i think it herokuish in sudoers to be exact, but didn't do much testing cuz it works

Thanks a lot for your patience :)
